chaos/cluster/cluster.go

```diff
@@ -0,0 +1,219 @@
+
+// RebootAll reboots all the nodes that are accessible ie. have ExternalIP.
+// rebootAfter is the time duration between stopping kubelet and issuing reboot.
+func (cl *Cluster) RebootAll(rebootAfter time.Duration) error {
+	var hosts []string
+	hosts = append(hosts, hostsFromNodes(cl.Masters)...)
+	hosts = append(hosts, hostsFromNodes(cl.Workers)...)
+	if len(hosts) < 1 {
+		return fmt.Errorf("no nodes found that can be rebooted")
+	}
+
+	glog.V(4).Infof("will reboot nodes: %s", hosts)
+	maxParallel, err := intstr.GetValueFromIntOrPercent(&cl.MaxDisruption, len(hosts), true)
+	if err != nil {
+		return fmt.Errorf("errors parsing max disruption: %v", err)
+	}
+
+	return cl.rebootHosts(hosts, maxParallel, rebootAfter)
+}
+
+// RebootAllMasters reboots all the master nodes that are accessible ie. have ExternalIP.
+// rebootAfter is the time duration between stopping kubelet and issuing reboot.
+func (cl *Cluster) RebootAllMasters(rebootAfter time.Duration) error {
+	hosts := hostsFromNodes(cl.Masters)
+	if len(hosts) < 1 {
+		return fmt.Errorf("no nodes found that can be rebooted")
+	}
+
+	glog.V(4).Infof("will reboot nodes: %s", hosts)
+	maxParallel, err := intstr.GetValueFromIntOrPercent(&cl.MaxDisruption, len(hosts), true)
+	if err != nil {
+		return fmt.Errorf("errors parsing max disruption: %v", err)
+	}
+
+	return cl.rebootHosts(hosts, maxParallel, rebootAfter)
+}
+
+// RebootAllWorkers reboots all the worker nodes that are accessible ie. have ExternalIP.
+// rebootAfter is the time duration between stopping kubelet and issuing reboot.
+func (cl *Cluster) RebootAllWorkers(rebootAfter time.Duration) error {
+	hosts := hostsFromNodes(cl.Workers)
+	if len(hosts) < 1 {
+		return fmt.Errorf("no nodes found that can be rebooted")
+	}
+
+	glog.V(4).Infof("will reboot nodes: %s", hosts)
+	maxParallel, err := intstr.GetValueFromIntOrPercent(&cl.MaxDisruption, len(hosts), true)
+	if err != nil {
+		return fmt.Errorf("errors parsing max disruption: %v", err)
+	}
+
+	return cl.rebootHosts(hosts, maxParallel, rebootAfter)
+}
+
+// RebootNode reboots a node addressable with `host`.
+// Uses *Cluster sshClient.
+// rebootAfter is the time duration between stopping kubelet and issuing reboot.
+func (cl *Cluster) RebootNode(host string, rebootAfter time.Duration) error {
+	cmd := cmdReboot
+	if rebootAfter.Seconds() > 0 {
+		cmd = fmt.Sprintf(cmdRebootWithSleep, int(rebootAfter.Seconds()))
+	}
+
+	glog.V(4).Infof("node: %s rebooting", host)
+	glog.V(4).Infof("node: %s executing cmd: '%s'", host, cmd)
+	stdout, stderr, err := cl.sshClient.Exec(host, cmd)
+	if _, ok := err.(*ssh.ExitMissingError); ok {
+		// A terminated session is perfectly normal during reboot.
+		err = nil
+	}
+	if err != nil {
+		return fmt.Errorf("node: %s issuing command failed\nstdout:%s\nstderr:%s", host, stdout, stderr)
+	}
+	glog.V(4).Infof("node: %s rebooted waiting for node to come back up", host)
+	if err := wait.PollImmediate(1*time.Second, 1*time.Minute, func() (bool, error) {
```

Comment:
which upstream?

usually, it is better to let one func do one thing rather than putting everything together.

3 steps == 3 funcs.